plugins/techdocs-backend/src/search/DefaultTechDocsDocumentGenerator.ts

```diff
@@ -0,0 +1,194 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  PluginEndpointDiscovery,
+  TokenManager,
+} from '@backstage/backend-common';
+import { Entity, RELATION_OWNED_BY } from '@backstage/catalog-model';
+import fetch from 'cross-fetch';
+import unescape from 'lodash/unescape';
+import { Logger } from 'winston';
+import pLimit from 'p-limit';
+import { Config } from '@backstage/config';
+import { CatalogApi, CatalogClient } from '@backstage/catalog-client';
+import { TechDocsDocument } from '@backstage/techdocs-common';
+
+interface MkSearchIndexDoc {
+  title: string;
+  text: string;
+  location: string;
+}
+
+export type TechDocsCollatorOptions = {
+  discovery: PluginEndpointDiscovery;
+  logger: Logger;
+  tokenManager: TokenManager;
+  locationTemplate?: string;
+  catalogClient?: CatalogApi;
+  parallelismLimit?: number;
+  legacyPathCasing?: boolean;
+};
+
+type EntityInfo = {
+  name: string;
+  namespace: string;
+  kind: string;
+};
+
+export class DefaultTechDocsDocumentGenerator {
+  private discovery: PluginEndpointDiscovery;
+  private locationTemplate: string;
+  private readonly logger: Logger;
+  private readonly tokenManager: TokenManager;
+  private readonly catalogClient: CatalogApi;
+  private readonly parallelismLimit: number;
+  private readonly legacyPathCasing: boolean;
+  public readonly type: string = 'techdocs';
+
+  private constructor({
+    discovery,
+    locationTemplate,
+    logger,
+    tokenManager,
+    catalogClient,
+    parallelismLimit = 10,
+    legacyPathCasing = false,
+  }: TechDocsCollatorOptions) {
+    this.discovery = discovery;
+    this.locationTemplate =
+      locationTemplate || '/docs/:namespace/:kind/:name/:path';
+    this.logger = logger;
+    this.tokenManager = tokenManager;
+    this.catalogClient =
+      catalogClient || new CatalogClient({ discoveryApi: discovery });
+    this.parallelismLimit = parallelismLimit;
+    this.legacyPathCasing = legacyPathCasing;
+  }
+
+  static fromConfig(config: Config, options: TechDocsCollatorOptions) {
+    const legacyPathCasing =
+      config.getOptionalBoolean(
+        'techdocs.legacyUseCaseSensitiveTripletPaths',
+      ) || false;
+    return new DefaultTechDocsDocumentGenerator({
+      ...options,
+      legacyPathCasing,
+    });
+  }
+
+  async *execute(): AsyncGenerator<TechDocsDocument, void, undefined> {
+    const limit = pLimit(this.parallelismLimit);
+    const techDocsBaseUrl = await this.discovery.getBaseUrl('techdocs');
+    const { token } = await this.tokenManager.getToken();
+    const entities = await this.catalogClient.getEntities(
```

Comment:
At this point all the entities are still in memory. I'm missing where the memory's optimization is 😅

Reply:
Of catalog entities! But I suspect the real memory hog in affected systems is TechDocs _content_ (which is a little over 2x the actual size of the text content due to what/how mkdocs produces it's search indices)

Though yes, it'd be great to make the improvements needed to the catalog API/client to stream or paginate through entities too.

Reply:
The catalog API does have pagination (on the server side), but the client lib doesn't have the input parameters yet to expose it 🤔

Reply:
Alright, added. https://github.com/backstage/backstage/pull/8507

Reply:
And now, having rebased, we're taking advantage of this now.